text/0340-firehose-l2.md

```diff
@@ -0,0 +1,944 @@
+const s3Destination = new destinations.S3({
+  compression: Compression.SNAPPY,
+  bucket: bucket,
+});
+new DeliveryStream(this, 'Delivery Stream', {
+  destination: destination,
+});
+```
+
+## Buffering
+
+Incoming data is buffered before it is delivered to the specified destination. The
+delivery stream will wait until the amount of incoming data has exceeded some threshold
+(the "buffer size") or until the time since the last data delivery occurred exceeds some
+threshold (the "buffer interval"), whichever happens first. You can configure these
+thresholds based on the capabilities of the destination and your use-case. By default, the
+buffer size is 3 MiB and the buffer interval is 1 minute.
+
+```ts fixture=with-bucket
+// Increase the buffer interval and size to 5 minutes and 3 MiB, respectively
+import * as cdk from '@aws-cdk/core';
+
+const s3Destination = new destinations.S3({
+  bufferingInterval: cdk.Duration.minutes(5),
+  bufferingSize: cdk.Size.mebibytes(8),
+  bucket: bucket,
+});
+new DeliveryStream(this, 'Delivery Stream', {
+  destination: destination,
+});
+```
+
+See: [Data Delivery Frequency](https://docs.aws.amazon.com/firehose/latest/dev/basic-deliver.html#frequency)
+in the *Kinesis Data Firehose Developer Guide*.
+
+## Backup
+
+A delivery stream can be configured to backup data to S3 that it attempted to deliver to
+the configured destination. Backed up data can be all the data that the delivery stream
+attempted to deliver or just data that it failed to deliver (Redshift and S3 destinations
+can only backup all data). CDK can create a new S3 bucket where it will back up data or
+you can provide a bucket where data will be backed up. You can also provide prefix under
+which your backed-up data will placed within the bucket. By default, source data is not
+backed up to S3.
+
+```ts fixture=with-domain
+import * as destinations from '@aws-cdk/aws-kinesisfirehose-destinations';
+import * as s3 from '@aws-cdk/aws-s3';
+
+// Enable backup of all source records (to an S3 bucket created by CDK)
+const deliveryStream = new DeliveryStream(this, 'Delivery Stream Backup All', {
+  destination: new destinations.Elasticsearch({
+    domain: domain,
+    indexName: 'myindex',
+    backup: BackupMode.ALL,
+  }),
+});
+
+// Enable backup of only the source records that failed to deliver (to an S3 bucket created by CDK)
+const deliveryStream = new DeliveryStream(this, 'Delivery Stream Backup Failed', {
+  destination: new destinations.Elasticsearch({
+    domain: domain,
+    indexName: 'myindex',
+    backup: BackupMode.FAILED,
+  }),
+});
+
+// Explicitly provide an S3 bucket to which all source records will be backed up
+const backupBucket = new s3.Bucket(this, 'Bucket');
+const deliveryStream = new DeliveryStream(this, 'Delivery Stream Backup All Explicit Bucket', {
+  destination: new destinations.Elasticsearch({
+    domain: domain,
+    indexName: 'myindex',
+    backupBucket: backupBucket,
```

Comment:
S3 backup is supported by all destinations right now. So we could move it up to the stream instead of the destination.

The Cloudformation API design has the S3 backup info as part of the destination, not part of the stream. But not sure, if that is a good reason to keep the S3 backup as part of the destination. 

We've also reached out to the Firehose team to try to find out if they will support multiple destinations in the future. But you are right, it is only one right now.

Reply:
As you point out Rico, I think we made a lot of the decisions about where to place configuration based on the fact that the API seems to be leaning towards supporting multiple destinations. You'll have to let us know if that's unnecessary future-proofing or a smart design decision. Hopefully the service team can get back to us about their future directions

Reply:
As above, we will be supporting multiple destinations